Add a quickstart tutorial (#103)
cruise-automation/webviz: docs/src/2.5.AddingInteractivity.mdx

```js
const [clickedObjectIds, setClickedObjectIds] = useState(new Set());

const obstacleMarkers = Array.from(clickedObjectIds).map((clickedObjectId, index) => {
  const pointIdx = clickedObjectId - marker.id;
  const position = marker.points[pointIdx];
  return {
    // keep the id unique by adding sphere's id range as offset: marker.id + steps
    id: marker.id + steps + index,
    // remember the original clickedObjectId so when the obstacle is clicked, we can
    // remove the obstacle quickly by updating clickedObjectIds
    clickedObjectId,
    pose: {
      orientation: { x: 0, y: 0, z: 0, w: 1 },
      position: position,
    },
    color: { r: 1, g: 0, b: 0, a: 1 }, // red
    scale: { x: 4, y: 4, z: 4 }, // scale up a little so it's bigger than the spheres
  };
});
```

Now, let's use the clicked sphere id to update the `clickedObjectIds` state, which will be consumed by the cube obstacles. Clicking each obstacle should remove it from the scene.

```js
 <Spheres
    onClick={(ev, { objectId, object }) => {
        // click on the object to add an obstacle
        clickedObjectIds.add(objectId);

>         setClickedObjectIds(new Set(clickedObjectIds));

Reviewer: Seems pointless to use `.add()` and then `new Set()`. At least we could use `setClickedObjectIds(clickedObjectIds.add(objectId))` (assuming that will actually cause a re-render, which I think it will). Alternatively, could just use an array and then `setClickedObjectIds([...clickedObjectIds, objectId])` for full immutability (although it's a bit less efficient, but that doesn't matter much because the number of objects will be quite small)

    }}>
    {[marker]}
    </Spheres>
    <Cubes
        onClick={(ev, { objectId, object }) => {

>             // the `object` contains information of the clicked cube, which has `clickedObjectId`

Reviewer: Needs more explanation why you can't just use `objectId`

            clickedObjectIds.delete(object.clickedObjectId);
            setClickedObjectIds(new Set(clickedObjectIds));
        }}>
        {obstacleMarkers}
    </Cubes>
```

> Lastly, setting an `shouldStopDuck` variable will allow us to tell the duck when to stop and when to move forward.

Reviewer: ```suggestion
Lastly, setting a `shouldStopDuck` variable will allow us to tell the duck when to stop and when to move forward.
```

```js

> // keep track whethere we need to stop the duck from moving

Reviewer: ```suggestion
// keep track of whether we need to stop the duck from moving
```

Consider flipping this, so it's `shouldMove` or something

let shouldStopDuck = false;
useRequestAnimationFrame(() => {

>   if (!shouldStopDuck) {

Reviewer: Your callback does nothing if this variable is false. Consider making this bool a *parameter to useRequestAnimationFrame* so you can actually stop the loop while the duck is not moving. Similar to what I did in useEventListener:

https://github.com/cruise-automation/webviz/blob/72bdc97f1d00ef7cb8832e7c6b615bc8d4392ad4/docs/src/jsx/utils/useEventListener.js#L12

    // update count before each browser repaint

>     const newCount = count >= 100000 ? 0 : count + 1;

Reviewer: nit: `(count + 1) % 100000`  (but see comment above about requestAnimationFrame timestamp)

    setCount(newCount);
  }
});

obstacleMarkers.forEach((marker) => {
  const position = marker.pose.position;
  // make the duck stop when the duck's current position has an obstacle
  if (position.x === duckPosition.x && position.y === duckPosition.y && position.z === duckPosition.z) {
    shouldStopDuck = true;
  }
});
s;
```

> Putting it all together, we now have a moving duck that gets stuck on every cube obstacle in its path. Clicking a blocking cube will remove it, allowing the duck to happily continue forward.

Reviewer: Text is missing explanation of the fact that you can add a cube by clicking a sphere

<CodeSandboxEmbed src="https://codesandbox.io/s/j7n5r2mkr3" />

## Next Steps

> Congratulations on finishing this tutorial, and building your first `regl-worldview` app! To learn more:

Reviewer: ```suggestion
Congratulations on finishing this tutorial, and building your first app using Worldview! To learn more:
```

- Read the [API docs](#/docs/api) to learn common APIs for Worldview, the camera, mouse interaction, type support, etc.

> - Read about supported WebGL [Commands](#/docs/commands) – `Cubes`, `Spheres`, `Triangles`, `Lines`, etc.

Reviewer: Maybe also want to point them to GitHub for issues/PRs? Point to regl documentation?

Author: Good idea! Added

import CodeSandboxEmbed from './jsx/utils/CodeSandboxEmbed.js';

# Adding Interactivity

In this page, you'll learn how to:

- set up the marker input for hitmap/mouse interaction
- interact with the scene through event handlers
- retrive object and event information from the event handlers

## Adding and Removing Obstacles along the Duck's Path

In previous steps, you've learned how to interact with Worldview through mouse, keyboard, and programatic control, all of which affect the camera. Worldiview and the Commmand components also support other mouse interactions including `onClick`, `onDoubleClick`, `onMouseDown`, `onMouseMove`, and `onMouseUp`. Learn more at the [Mouse Events](#/docs/api/mouse-events) page.

> In this section, we are going to place obstacles (red cubes) along the duck's moving path, and when the duck hit the obstacles, it's going to stop moving until you removed the blocking obstacles.

Reviewer 2: 'and when the duck hit**s** the obstacles'

First, let's add the `id` field to the sphere marker to enable hitmap so we correctly identify the specific sphere that's being clicked.

```js
const marker = {
  // hitmap support requires unique positive integer id
  id: 101,
  pose: {
    orientation: { x: 0, y: 0, z: 0, w: 1 },
    position: { x: 0, y: 0, z: 0 },
  },
  scale: { x: 1, y: 1, z: 1 },
  colors: [],
  points: [],
};
```

Let's also enable manual camera movement, so we can see the duck better.

```js
<Worldview defaultCameraState={{ ...DEFAULT_CAMERA_STATE, distance: 120 }}>
```

Next, we are going to add a `clickedObjectIds` state to keep track of the clicked spheres. We'll use it to generate obstacles which will be reprensented by big red cubes.

```js
const [clickedObjectIds, setClickedObjectIds] = useState(new Set());

const obstacleMarkers = Array.from(clickedObjectIds).map((clickedObjectId, index) => {
  const pointIdx = clickedObjectId - marker.id;
  const position = marker.points[pointIdx];
  return {

>     //keep the id unique by adding sphere's id range as offset: marker.id + steps

Reviewer 2: nit space after `//`

    id: marker.id + steps + index,
    // remember the original clickedObjectId so when the obstacle is clicked, we can

>     // remove the obstacle quickly by update clickedObjectIds

Reviewer 2: 'by **updating** clickedObjectIds'

    clickedObjectId,
    pose: {
      orientation: { x: 0, y: 0, z: 0, w: 1 },
      position: position,
    },
    color: { r: 1, g: 0, b: 0, a: 1 }, // red
    scale: { x: 4, y: 4, z: 4 }, // scale up a little so it's bigger than the spheres
  };
});
```

The next step is to get the clicked sphere id and update `clickedObjectIds` state which will be consumed by the red cube obstacles. The user can click each obstacle to remove it.

```js
 <Spheres
    onClick={(ev, { objectId, object }) => {
        // Click once to add obstacle. Even though both the small Sphere and the big Obstacle cube have

>         // hitmap enabled, clicking again will only trigger `onClick` event on the Cube because it's a lot bigger

Reviewer 2: I don't know if the 2nd comment is necessary

        clickedObjectIds.add(objectId);
        setClickedObjectIds(new Set(clickedObjectIds));
    }}>
    {[marker]}
    </Spheres>
    <Cubes
        onClick={(ev, { objectId, object }) => {
            // the `object` contains information of the clicked Cube, which has `clickedObjectId`
            clickedObjectIds.delete(object.clickedObjectId);
            setClickedObjectIds(new Set(clickedObjectIds));
        }}>
        {obstacleMarkers}
    </Cubes>
```

Lastly, we need a way to tell the duck to stop or move foward. A simple variable `stopDuckMoving` is enough:

```js
// keep track whethere we need to stop the duck from moving
let stopDuckMoving = false;
useRequestAnimationFrame(() => {
  if (!stopDuckMoving) {
    // update count before each browser repaint
    const newCount = count >= 100000 ? 0 : count + 1;
    setCount(newCount);
  }
});

obstacleMarkers.forEach((marker) => {
  const position = marker.pose.position;
  // make the duck stop when the duck's current position has an obstacle
  if (position.x === duckPosition.x && position.y === duckPosition.y && position.z === duckPosition.z) {
    stopDuckMoving = true;
  }
});
s;
```

> Put it all together, we now have a moving duck that get stuck by every red cube obstacles we put on it's path. Click the cube that the duck is stuck in to remove it and the duck will continue moving forward.

Reviewer 2: 'Putting it all together, we now have a moving duck that gets stuck by every red cube obstacle we put on its path.'
'Click the cube that the duck is stuck in to remove it and the duck will continue moving forward.'

<CodeSandboxEmbed src="https://codesandbox.io/s/j7n5r2mkr3" />

## Next Steps

> Congratulations on finishing this tutorial, and building your first  `regl-worldview` app! To learn more:

Reviewer 3: Oops – this extra space might have been me.

Author: not sure there is any extra space...

import CodeSandboxEmbed from './jsx/utils/CodeSandboxEmbed.js';

# Adding Interactivity

In this chapter, you'll learn how to:

- Set up the objects in the scene for mouse interaction
- Interact with objects through event handlers
- Retrieve object and event information from the event handlers

## Adding and Removing Obstacles along the Duck's Path

In previous chapters, you've learned how to interact with Worldview through mouse, keyboard, and programmatic controls, all of which affect the camera. Worldview and its [Command components](#/docs/api/command) also support other [mouse events](#/docs/api/mouse-events), such as `onClick`, `onDoubleClick`, `onMouseDown`, `onMouseMove`, and `onMouseUp`.

In this step, let's make the spheres turn into large cubes when clicked. The duck will travel until it hits one of the cubes, and stop moving until you click again on the obstacle to remove it.

> First, let's add an `id` field to each sphere so that we can identify which one was clicked. The `id` value must be unique positive integers.

Reviewer 4: ```suggestion
First, let's add an `id` field to each sphere so that we can identify which one was clicked. The `id` value must be a unique positive integer.
```

```js
const sphereMarker = {
  id: 101,
  pose: {
    orientation: { x: 0, y: 0, z: 0, w: 1 },
    position: { x: 0, y: 0, z: 0 },
  },
  scale: { x: 2, y: 2, z: 2 }, // scale up so it's easy to click
  colors: [],
  points: [],
};
```

Let's also enable manual camera movement, so we can see the duck better.

```js

> // use `defaultCameraState` to turn Worldview to uncontrolled component

Reviewer 4: ```suggestion
// use `defaultCameraState` to turn Worldview into an uncontrolled component
```
#nit
